commonmark/src/main/java/org/commonmark/internal/DocumentParser.java

```diff
@@ -1,14 +1,14 @@
 package org.commonmark.internal;
 
+import java.io.IOException;
 import org.commonmark.internal.util.Parsing;
-import org.commonmark.internal.util.Substring;
 import org.commonmark.node.*;
 import org.commonmark.parser.block.*;
 
 import java.util.*;
 
-public class DocumentParser implements ParserState {
-
+public abstract class DocumentParser<T> implements ParserState {
```

Comment:
I think adding a class hierarchy for this is a bit too complicated. I'd prefer to just have another method on `DocumentParser` that takes a `Reader`. The code before `incorporateLine` can be moved into the constructor, the part after into a single method that needs to be called by both `parse` methods.